File: modules/motocortex-obd/ios/session/OBDSessionManager.swift
import Foundation

class OBDSessionManager {
    enum ConnectionState: String {
        case disconnected = "DISCONNECTED"
        case connecting = "CONNECTING"
        case connected = "CONNECTED"
        case reconnecting = "RECONNECTING"
    }

    private(set) var currentState = ConnectionState.disconnected {
        didSet {
            onStateChangedListener?(currentState.rawValue)
        }
    }

    private var transport: OBDTransport?
    private(set) var queue: OBDCommandQueue?

    private var onStateChangedListener: ((String) -> Void)?
    private var telemetryListener: ((String) -> Void)?

    private let sessionQueue = DispatchQueue(label: "com.motocortex.obd.session")
    private var heartbeatTimer: Timer?
    private var lastActivityTime = Date()

    private var currentType: String?
    private var currentTarget: String?

    func setOnStateChangedListener(listener: @escaping (String) -> Void) {
        onStateChangedListener = listener
    }

    func setTelemetryListener(listener: @escaping (String) -> Void) {
        telemetryListener = listener
    }

    func connect(type: String, target: String, completion: @escaping (Bool) -> Void) {
        currentType = type
        currentTarget = target
        currentState = .connecting

        let newTransport: OBDTransport
        switch type.lowercased() {
        case "ble":
            newTransport = BLETransport()
        case "wifi":
            newTransport = WiFiTransport()
        default:
            currentState = .disconnected
            completion(false)
            return
        }

        transport = newTransport
        let obdQueue = OBDCommandQueue(transport: newTransport)
        queue = obdQueue

        newTransport.connect(target: target) { [weak self] success in
            guard let self = self else { return }
            if success {
                obdQueue.startProcessing()
                self.currentState = .connected
                self.startHeartbeat()
                self.setupDataLogging()
                completion(true)
            } else {
                self.currentState = .disconnected
                completion(false)
            }
        }
    }

    func disconnect() {
        stopHeartbeat()
        queue?.stopProcessing()
        transport?.disconnect()
        queue = nil
        transport = nil
        currentState = .disconnected
    }

    private func setupDataLogging() {
        transport?.setOnDataReceivedListener { [weak self] data in
            guard let self = self else { return }
            self.lastActivityTime = Date()
            self.telemetryListener?(data)
        }
    }

    /**
     * Cranking Recovery (Warm-Start):
     * Triggered on Broken Pipe / voltage drops during cranking.
     * Flushes queue and restarts socket within 500ms cooldown.
     */
    func triggerCrankingRecovery() {
        if currentState == .reconnecting { return }
        print("[OBDSessionManager] Cranking detected! Initiating 500ms Warm-Start...")
        currentState = .reconnecting
        
        // Immediately flush the queue (clean-state)
        queue?.flush()
        
        sessionQueue.asyncAfter(deadline: .now() + 0.5) { [weak self] in
            guard let self = self, let type = self.currentType, let target = self.currentTarget else { return }
            self.connect(type: type, target: target) { reconnected in
                if !reconnected {
                    print("[OBDSessionManager] Cranking Recovery failed. Retrying in 2 seconds...")
                    self.sessionQueue.asyncAfter(deadline: .now() + 2.0) {
                        if self.currentState == .reconnecting {
                            self.connect(type: type, target: target) { _ in }
                        }
                    }
                }
            }
        }
    }

    private func startHeartbeat() {
        stopHeartbeat()
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.heartbeatTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { _ in
                self.sessionQueue.async {
                    let now = Date()
                    if now.timeIntervalSince(self.lastActivityTime) >= 30.0 && self.currentState == .connected {
                        if let q = self.queue {
                            print("[OBDSessionManager] Sending Keep-Alive (AT RV)...")
                            q.enqueue(commandString: "AT RV", timeoutMs: 1000) { result in
                                switch result {
                                case .success:
                                    self.lastActivityTime = Date()
                                case .failure:
                                    self.triggerCrankingRecovery()
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    private func stopHeartbeat() {
        DispatchQueue.main.async { [weak self] in
            self?.heartbeatTimer?.invalidate()
            self?.heartbeatTimer = nil
        }
    }
}

File: modules/motocortex-obd/android/src/main/java/com/motocortex/obd/session/OBDSessionManager.kt
package com.motocortex.obd.session

import android.content.Context
import com.motocortex.obd.queue.OBDCommandQueue
import com.motocortex.obd.transport.BLETransport
import com.motocortex.obd.transport.ClassicTransport
import com.motocortex.obd.transport.WiFiTransport
import com.motocortex.obd.transport.OBDTransport
import kotlinx.coroutines.*

class OBDSessionManager(private val context: Context) {
    enum class ConnectionState {
        DISCONNECTED,
        CONNECTING,
        CONNECTED,
        RECONNECTING
    }

    var currentState = ConnectionState.DISCONNECTED
        private set(value) {
            field = value
            onStateChangedListener?.invoke(value.name)
        }

    private var transport: OBDTransport? = null
    var queue: OBDCommandQueue? = null
        private set

    private var onStateChangedListener: ((String) -> Unit)? = null
    private var telemetryListener: ((String) -> Unit)? = null

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var heartbeatJob: Job? = null
    private var lastActivityTime = System.currentTimeMillis()

    private var currentType: String? = null
    private var currentTarget: String? = null

    fun setOnStateChangedListener(listener: (String) -> Unit) {
        onStateChangedListener = listener
    }

    fun setTelemetryListener(listener: (String) -> Unit) {
        telemetryListener = listener
    }

    fun connect(type: String, target: String): Boolean = runBlocking {
        currentType = type
        currentTarget = target
        currentState = ConnectionState.CONNECTING

        val newTransport = when (type.lowercase()) {
            "bluetooth" -> ClassicTransport()
            "ble" -> BLETransport(context)
            "wifi" -> WiFiTransport()
            else -> {
                currentState = ConnectionState.DISCONNECTED
                return@runBlocking false
            }
        }

        transport = newTransport
        val obdQueue = OBDCommandQueue(newTransport)
        queue = obdQueue

        val connected = withContext(Dispatchers.IO) {
            newTransport.connect(target)
        }

        if (connected) {
            obdQueue.startProcessing()
            currentState = ConnectionState.CONNECTED
            startHeartbeat()
            setupDataLogging()
            true
        } else {
            currentState = ConnectionState.DISCONNECTED
            false
        }
    }

    fun disconnect() {
        stopHeartbeat()
        queue?.stopProcessing()
        transport?.disconnect()
        queue = null
        transport = null
        currentState = ConnectionState.DISCONNECTED
    }

    private fun setupDataLogging() {
        transport?.setOnDataReceivedListener { data ->
            lastActivityTime = System.currentTimeMillis()
            telemetryListener?.invoke(data)
        }
    }

    /**
     * Cranking Recovery (Warm-Start):
     * Triggered on Broken Pipe / voltage drops during cranking.
     * Flushes queue and restarts socket within 500ms cooldown.
     */
    fun triggerCrankingRecovery() {
        if (currentState == ConnectionState.RECONNECTING) return
        System.err.println("[OBDSessionManager] Cranking detected! Initiating 500ms Warm-Start...")
        currentState = ConnectionState.RECONNECTING
        
        // Immediately flush the queue to start with a clean state on reconnect
        queue?.flush()
        
        scope.launch {
            delay(500) // Cooldown 500ms before warm-starting socket
            val type = currentType ?: return@launch
            val target = currentTarget ?: return@launch
            
            val reconnected = connect(type, target)
            if (!reconnected) {
                System.err.println("[OBDSessionManager] Cranking Recovery failed. Retrying in 2 seconds...")
                delay(2000)
                if (currentState == ConnectionState.RECONNECTING) {
                    connect(type, target)
                }
            }
        }
    }

    private fun startHeartbeat() {
        stopHeartbeat()
        heartbeatJob = scope.launch {
            while (isActive) {
                delay(1000)
                val now = System.currentTimeMillis()
                // Heartbeat ping (keep-alive) if idle for 30 seconds
                if (now - lastActivityTime >= 30000 && currentState == ConnectionState.CONNECTED) {
                    val q = queue
                    if (q != null) {
                        try {
                            System.out.println("[OBDSessionManager] Sending Keep-Alive (AT RV)...")
                            q.enqueue("AT RV", 1000L).await()
                            lastActivityTime = System.currentTimeMillis()
                        } catch (e: Exception) {
                            // If heartbeat fails, initiate Cranking Recovery
                            triggerCrankingRecovery()
                        }
                    }
                }
            }
        }
    }

    private fun stopHeartbeat() {
        heartbeatJob?.cancel()
        heartbeatJob = null
    }
}
